### DISCUSSION TITLE
[question] How to type provided value in api through middleware (typescript)

### DISCUSSION THREAD

**Discussion by zecka** [pmndrs/zustand#1018] [Q&A] [answered]
Hello, I try to provide an helpers function to all my stores trough a middleware. 

But I can not find the way to **type the api object**.

Here is a codesandbox to highlight my issue: https://codesandbox.io/s/zustand-middleware-typescript-provide-method-60bz2v?file=/src/zustand-demo.ts

Here is my custom "createStore" who use my middleware

```ts
// zustand-helper.ts
export const createStore = <StoreType extends {}>(
  fn: StateCreator<StoreType>
) => {
  const customMiddleware: Middleware<StoreType> = config => (set, get, api) => {
    const test = () => {
      console.log("THIS IS A TEST")
    }
    return config(set, get, { ...api, test }, [])
  }
  return create(customMiddleware(fn))
}
```
Here I got a type error on test in config return:
![image](https://user-images.githubusercontent.com/18116930/174324834-e94041c9-b0ab-4294-a59c-ae436bb984a8.png)

And this how I use my store

```ts
// zustand-demo.ts
import { createStore } from "./zustand-helper"
type DemoStore = {
  test: string
  setTest: (test: string) => void
}
export const useDemoStore = createStore<DemoStore>((set, get, api) => ({
  test: "",
  setTest(value: string) {
    api.test()
    set({ test: value })
  }
}))
```
![image](https://user-images.githubusercontent.com/18116930/174325103-70949acc-b686-4636-8ff0-fbbab4683792.png)

**Comment by dai-shi** [comment, marked as answer]
Creating a custom middleware with TypeScript is a tough task. (In JS, it's easy.)
You need to read this carefully and understand the idea.
https://github.com/pmndrs/zustand/blob/main/docs/typescript.md#authoring-middlewares-and-advanced-usage

**Comment by zecka** [comment]
Thank you very much for your quick response, it is effectively a tough task 😅

I read documentation carefully but I’m lost. I have some lack of knowledge about the advanced use of typescript.

I’m currently trying to moving from [“Global hook store”](https://www.npmjs.com/package/global-hook-store) to Zustand.

 [“Global hook store”](https://www.npmjs.com/package/global-hook-store) provide something called “asyncState”. I want to replicate this behaviour on zustand.


  <h3> Async state have following interface:</h3>
  
 
```tsx
type AsyncState<T = any> = {
  error?: any;
  loading: boolean;
  data: T | null;
};
```

To avoid duplicate code, I would like to provide an helper to set an async state with a promise

Something with an interface like:

```ts

type KeyOfType<T, U> = {[P in keyof T]: T[P] extends U ? P: never}[keyof T]

type SetAsyncState = (
  stateKey: KeyOfType<State, AsyncState>,
  promise: Promise<StateValue>
) => Promise<boolean>

```
I did it but, i have to provide `set` and `get` function every time I call the helper.


<details>
 <summary>
 <h3>Here is my current helper:</h3>
 </summary>

```tsx
export const setAsyncState = async (
  stateKey: any,
  promise: Promise<any>,
  set: any,
  get: any
): Promise<boolean> => {
  const currentState: AsyncState = get()[stateKey]
  set({ [stateKey]: { ...currentState, loading: true } })
  try {
    const result = await promise
    set({ [stateKey]: { loading: false, data: result } })
    return true
  } catch (e) {
    set({ [stateKey]: { ...currentState, loading: false, error: e } })
    return false
  }
}
```

</details>
My goal is to ehance typescript experience with a middleware. To enforce typing validation and also avoid having to provide the get and set function.

I have so far created my middleware with a lot of `any`.


<details>
 <summary>
 <h3>Here is the current middleware ()</h3>
 </summary>

```tsx

type AsyncState<T = any> = {
  error?: any;
  loading: boolean;
  data: T | null;
};

type SetAsyncState<StateKey, StateValueType> = (
  stateKey: StateKey,
  actionPromise: Promise<StateValueType>
) => Promise<void>;


const asyncStateMiddlewareImpl: any = (f, bar) => (set, get, _store) => {
  const store = _store as any;

  const setAsyncState: SetAsyncState<any, any> = async (
    stateKey,
    actionPromise
  ) => {
    const myCurrentState: AsyncState = get()[stateKey];
    set({
      [stateKey]: {
        ...myCurrentState,
        loading: true
      }
    });
    try {
      const result = await actionPromise;
      set({
        [stateKey]: {
          loading: false,
          data: result
        }
      });
    } catch (e) {
      set({
        [stateKey]: {
          ...myCurrentState,
          loading: false,
          error: e
        }
      });
    }
  };

  store.setAsyncState = setAsyncState;

  return f(set, get, _store);
};

export const asyncStateMiddleware = (asyncStateMiddlewareImpl as unknown) as any;

```

</details>

(Or check it on codesandbox: https://codesandbox.io/s/zustand-try-to-implement-asyncstatemiddleware-ds5l8g)

### Is there any chance that someone can help me to type this helper?

**Comment by dai-shi** [comment]
Alternatively, because typing middleware is too tough, we can patch store after creation.

```js
const useStore = create(...)
useStore.setAsyncState = ...
```

**Comment by zecka** [comment]
Finally, I found another approach that seems to work well and has strong typing.

```ts
type DemoStore = {
  randomNumber: AsyncState<number>;
  loadRandomNumber: () => Promise<void>;
};
export const useStore = create<DemoStore>((set, get) => {
  const setAsyncState = useAsyncState<DemoStore>(set, get);
  return {
    randomNumber: defineAsyncState(2),
    classic: 1,
    loadRandomNumber: async () => {
      await setAsyncState("randomNumber", fakeFetch());
    }
  };
});


```
<details>
 <summary>
 <h3>And there is the async state helper</h3>
 </summary>

```tsx
export const asyncState = <T extends any>(defaultValue: T): AsyncState<T> => {
  return {
    loading: false,
    data: defaultValue
  }
}

export type AsyncState<T = any> = {
  error?: any
  loading: boolean
  data: T | null
}

export const setAsyncState = async (
  stateKey: any,
  promise: Promise<any>,
  set: any,
  get: any
): Promise<boolean> => {
  const currentState: AsyncState = get()[stateKey]
  set({ [stateKey]: { ...currentState, loading: true } })
  try {
    const result = await promise
    set({ [stateKey]: { loading: false, data: result } })
    return true
  } catch (e) {
    set({ [stateKey]: { ...currentState, loading: false, error: e } })
    return false
  }
}

type KeyOfType<T, U> = { [P in keyof T]: T[P] extends U ? P : never }[keyof T]

export const useAsyncState = <T = object>(set: any, get: any) => {
  return async <Key extends KeyOfType<T, AsyncState>>(
    stateKey: Key,
    promise: T[Key] extends AsyncState ? Promise<T[Key]["data"]> : never
  ) => {
    const currentState: AsyncState = get()[stateKey]

    set({ [stateKey]: { ...currentState, loading: true } })

    try {
      const result = await promise
      set({ [stateKey]: { loading: false, data: result } })
      return true
    } catch (e) {
      set({ [stateKey]: { ...currentState, loading: false, error: e } })
      return false
    }
  }
  return setAsyncState
}

```

</details>


## See it on codsandbox: https://codesandbox.io/s/zustand-use-async-state-zp9rji
